File: fsm/fsm.go
package main
 
import "fmt"
 
const (
	Start	= 1
	Update	= 0
	Stop	= -1
)
 
type StateFunc func(m *Machine, s int)
 
func empty(m *Machine, s int) {
}
 
type Machine struct {
	cur	StateFunc
	Next	StateFunc
}
 
func New() *Machine {
	return &Machine{empty, nil}
}
 
func (m *Machine) Update() {
	if m.Next != nil {
		m.cur(m, Stop)
		m.cur = m.Next
		m.Next = nil
		m.cur(m, Start)
	}
	m.cur(m, Update)
}
 
func tick(m *Machine, s int) {
	switch s {
	case Update:
		fmt.Println("tick!")
		m.Next = tock
	}
}
 
func tock(m *Machine, s int) {
	switch s {
	case Update:
		fmt.Println("tock!")
		m.Next = tick
	}
}
 
func main() {
	m := New()
	m.Next = tick
	for i := 0; i < 10; i += 1 {
		m.Update()
	}
}
 

File: fsm/fsm.js
state = {
	start : 1,
	update : 0,
	stop, -1
};

function Machine(){
	this.cur  = function(m,s){};
	this.next = null;
}

Machine.prototype.update = function(){
	if( this.next != null ){
		this.cur(this, state.stop);
		this.cur = this.next;
		this.next = null;
		this.cur(this, state.start);
	}
	this.cur(this, state.update);
};

// example
M = new Machine();

tick = function(m,s){
	switch(s){
		case state.start : 
			console.log("starting tick;");
			break;
		case state.update : 
			console.log("updating tick;");
			m.next = tock;
			break;
		case state.stop : 
			console.log("stopping tick;");
			break;
	}
};

tock = function(m,s){
	switch(s){
		case state.start : 
			console.log("starting tock;");
			break;
		case state.update : 
			console.log("updating tock;");
			m.next = tick;
			break;
		case state.stop : 
			console.log("stopping tock;");
			break;
	}
};

M.next = tick;

for(var i = 0; i < 100; i += 1){
	M.update();
}